packages/editor/src/optimizer/worker.ts

```diff
@@ -0,0 +1,366 @@
+ * `penroseState!`.
+ */
+let penroseState: PenroseState | null = null;
+let optimizerShouldStop = false;
+
+/**
+ * Returns a copy of parents varying values (or if null, a copy of the current).
+ * @param stepSequenceInfo
+ */
+const getParentOrCurrentVaryingValues = (
+  stepSequenceInfo: StepSequenceInfo,
+) => {
+  const parentLoc = stepSequenceInfo.parent;
+  return parentLoc === null
+    ? [...penroseState!.varyingValues]
+    : [...historyValues.get(parentLoc.sequenceId)![parentLoc.step]];
+};
+
+/**
+ * Compile, but don't start optimizing until we receive `LabelMeasurementsData`
+ * in a notification. After calling, `penroseState` will be non-null.
+ * @param data
+ */
+const compile = async (data: CompileRequestData): Promise<CompileResult> => {
+  const compiledState = await compileTrio(data);
+
+  let result: CompileResult;
+  if (Result.isOk(compiledState)) {
+    log.info("Compiled");
+    penroseState = compiledState.value;
+    makeNewStepSequence(null, data.variation);
+    // diagramId is assigned by broker, not us, so we just pass 0
+    result = taggedOk(MessageTags.Compile, {
+      diagramId: 0,
+      warnings: compiledState.value.warnings,
+    });
+  } else {
+    log.info("Compilation failed");
+    result = taggedErr(MessageTags.Compile, compiledState.error);
+  }
+
+  return result;
+};
+
+const poll = async (): Promise<PollResult> => {
+  return taggedOk(MessageTags.Poll, historyInfo.map());
+};
+
+const computeLayout = async (
+  data: ComputeLayoutRequestData,
+): Promise<ComputeLayoutResult> => {
+  const varyingValuesSequence = historyValues.get(data.historyLoc.sequenceId);
+  const stepSequenceInfo = historyInfo.get(data.historyLoc.sequenceId);
+
+  let result: ComputeLayoutResult;
+  if (
+    varyingValuesSequence &&
+    stepSequenceInfo &&
+    0 <= data.historyLoc.step &&
+    data.historyLoc.step < varyingValuesSequence.length
+  ) {
+    const xs = varyingValuesSequence[data.historyLoc.step];
+    result = taggedOk(
+      MessageTags.ComputeLayout,
+      stateToLayoutState({
+        ...penroseState!,
+        varyingValues: xs,
+      }),
+    );
+  } else {
+    result = taggedErr(MessageTags.ComputeLayout, {
+      tag: "InvalidHistoryLocError",
+      historyLoc: data.historyLoc,
+      historyInfo: historyInfo.map(),
+    });
+  }
+
+  return result;
+};
+
+const makeNewStepSequence = (
+  parentLoc: HistoryLoc | null,
+  variation: string,
+): StepSequenceID => {
+  const id: StepSequenceID = stepSequenceIdGenerator.next();
+
+  const info: StepSequenceInfo = {
+    layoutStats: [
+      {
+        name:
+          penroseState!.optStages.length === 1
+            ? "default"
+            : penroseState!.optStages[0],
+        // steps, as it is used in the rest of the codebase, indicates the number
+        // of varying values in a list--not the number of optimizer steps. So
+        // we initialize this to one, since one varying value is added initially
```

Comment:
Would it make sense to use a different word here, then?

Reply:
Sure, I can change it to "frame"; pretty sure this only affects `LayoutTimelineSlider` so shouldn't be a big change.